modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManagerImpl.java

```diff
@@ -17,53 +17,282 @@
 
     /** Configuration manager. */
     private final ConfigurationManager configurationMgr;
 
-    /** Raft manager. */
-    private final Loza raftMgr;
+    /** Table creation subscription future. */
+    private CompletableFuture<Long> tableCreationSubscriptionFut;
+
+    /** Tables. */
+    private Map<String, Table> tables;
 
     /**
-     * The constructor.
-     *
-     * @param configurationMgr Configuration table.
-     * @param clusterNetSvc Cluster network service.
-     * @param metaStorageMgr MetaStorage manager.
-     * @param schemaMgr Schema manager.
+     * @param configurationMgr Configuration manager.
+     * @param metaStorageMgr Meta storage manager.
+     * @param schemaManager Schema manager.
      * @param raftMgr Raft manager.
      */
     public TableManagerImpl(
         ConfigurationManager configurationMgr,
-        ClusterService clusterNetSvc,
         MetaStorageManager metaStorageMgr,
-        SchemaManager schemaMgr,
+        SchemaManager schemaManager,
         Loza raftMgr
     ) {
+        tables = new HashMap<>();
+
         this.configurationMgr = configurationMgr;
-        this.clusterNetSvc = clusterNetSvc;
         this.metaStorageMgr = metaStorageMgr;
-        this.schemaMgr = schemaMgr;
-        this.raftMgr = raftMgr;
+
+        String localNodeName = configurationMgr.configurationRegistry().getConfiguration(NetworkConfiguration.KEY)
+            .name().value();
+
+        configurationMgr.configurationRegistry().getConfiguration(LocalConfiguration.KEY)
+            .metastorageMembers().listen(ctx -> {
+            if (ctx.newValue() != null) {
+                if (hasMetastorageLocally(localNodeName, ctx.newValue()))
+                    subscribeForTableCreation();
+                else
+                    unsubscribeForTableCreation();
+            }
+            return CompletableFuture.completedFuture(null);
+
+        });
+
+        String[] metastorageMembers = configurationMgr.configurationRegistry().getConfiguration(LocalConfiguration.KEY)
+            .metastorageMembers().value();
+
+        if (hasMetastorageLocally(localNodeName, metastorageMembers))
+            subscribeForTableCreation();
+
+        String tableInternalPrefix = INTERNAL_PREFIX + "assignment.#";
+
+        tableCreationSubscriptionFut = metaStorageMgr.registerWatch(new Key(tableInternalPrefix), new WatchListener() {
+            @Override public boolean onUpdate(@NotNull Iterable<WatchEvent> events) {
+                for (WatchEvent evt : events) {
+                    if (!ArrayUtils.empty(evt.newEntry().value())) {
+                        String keyTail = evt.newEntry().key().toString().substring(INTERNAL_PREFIX.length());
+
+                        String placeholderValue = keyTail.substring(0, keyTail.indexOf('.'));
+
+                        UUID tblId = UUID.fromString(placeholderValue);
+
+                        try {
+                            String name = new String(metaStorageMgr.get(
```

Comment:
As were discussed please either use Vault or add TODO. See AffinityManager for more details.

Reply:
I have added a vault invokation.